lametro/views.py

```diff
@@ -734,6 +736,100 @@ class LAMetroArchiveSearch(TemplateView):
 class LAMetroContactView(IndexView):
     template_name = 'lametro/contact.html'
 
+
+class MinutesView(EventsView):
+    template_name = 'lametro/minutes.html'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        start_date_str = None
+        end_date_str = None
+        if 'minutes-from' in self.request.GET.keys():
+            start_date_str = self.request.GET.get('minutes-from')
+            start_datetime = datetime.datetime.strptime(start_date_str, '%m/%d/%Y')
+        if 'minutes-to' in self.request.GET.keys():
+            end_date_str = self.request.GET.get('minutes-to')
+            end_datetime = datetime.datetime.strptime(end_date_str, '%m/%d/%Y')
+
+        # filter db events based on start and end dates
+        if end_date_str:
+            context['end_date'] = end_date_str
+            end_date_time = parser.parse(end_date_str)
+            stored_events = LAMetroEvent.objects.with_media()\
+                                                .filter(start_time__lt=end_date_time)\
+                                                .order_by('start_time')
+        else:
+            stored_events = LAMetroEvent.objects.with_media()\
+                                                .filter(start_time__lt=timezone.now())\
+                                                .order_by('start_time')
+        if start_date_str:
+            context['start_date'] = start_date_str
+            start_date_time = parser.parse(start_date_str)
+            stored_events = stored_events.filter(start_time__gt=start_date_time)\
+                                                .order_by('start_time')
+
+        filtered_db_events = []
+        for event in stored_events:
+            stored_events_dict = {
+                'start_time': event.start_time.date(),
+                'meeting': event.name,
+                'minutes_link': [],
+                'agenda_link': [],
+            }
+            try:
+                minutes_link = event.documents.get(note__icontains='minutes').links.first().url
+                stored_events_dict['minutes_link'].append(minutes_link)
+            except EventDocument.DoesNotExist:
+                pass
+
+            try:
+                agenda_link = event.documents.get(note__icontains='agenda').links.first().url
+                stored_events_dict['agenda_link'].append(agenda_link)
+            except EventDocument.DoesNotExist:
+                pass
```

Comment:
This loop is going to execute a query for every event in the database, which will add up really quickly! I think this would be a good use case for [the raw method](https://docs.djangoproject.com/en/4.0/topics/db/sql/#adding-annotations):

```python
        # Write a query to retrieve the objects you want, plus any annotations (in this case, minutes and agenda links)
        query = '''
            with agenda_links as (
                select
                    ed.event_id,
                    edl.url
                from opencivicdata_eventdocument ed
                join opencivicdata_eventdocumentlink edl
                on ed.id = edl.document_id
                where ed.note ilike '%%agenda%%'
            ), minutes_links as (
                select
                    ed.event_id,
                    edl.url
                from opencivicdata_eventdocument ed
                join opencivicdata_eventdocumentlink edl
                on ed.id = edl.document_id
                where ed.note ilike '%%minutes%%'
            )
            select
                ce.event_id,
                e.name,
                e.start_date,
                al.url as agenda_link,
                ml.url as minutes_link
            from opencivicdata_event e
            join councilmatic_core_event ce
            on e.id = ce.event_id
            left join agenda_links al
            on e.id = al.event_id
            left join minutes_links ml
            on e.id = ml.event_id
            where ce.event_id in %s
        '''

        # Execute the query, passing in a tuple of event IDs within the specified time frame as a value for the filter parameter
        stored_events = LAMetroEvent.objects.raw(
            query, [tuple(stored_events.values_list('event_id', flat=True))]
        )
```

This approach will reduce the number of queries from the number of events in the database to 2: one to get the IDs of events within the time range, and another to get the document links for those events.

Reply:
try a prefetch method before raw method:

https://github.com/datamade/la-metro-councilmatic/pull/406